### PR TITLE
remove(demux post processing)

### PR DESCRIPTION
**Pull request by ChrOertlin**
## Description

removes demultiplex post processing api.

### Added

-

### Changed

- `cg demultiplex finish flow-cell` -> `cg demultiplex finish illumina-runs`

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/apps/demultiplex/sample_sheet/utils.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+LOG = logging.getLogger(__name__)
+
+
+def add_and_include_sample_sheet_path_to_housekeeper(
+    flow_cell_directory: Path, flow_cell_name: str, hk_api: HousekeeperAPI
+) -> None:
+    """
+    Add sample sheet path to Housekeeper.
+    Raises:
```
Comment (diitaz93) [inline cg/apps/demultiplex/sample_sheet/utils.py:19]: this function does not raise an error

File: cg/apps/demultiplex/sample_sheet/utils.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
+"""Utils for sample sheet generation."""
+
+import logging
+from pathlib import Path
+
```
Comment (diitaz93) [inline cg/apps/demultiplex/sample_sheet/utils.py:5]: For the type hint
```suggestion
from housekeeper.store.models import File
```

File: cg/cli/demultiplex/copy_novaseqx_demultiplex_data.py [hunk 3]
```diff
@@ -114,3 +117,103 @@ def is_ready_for_post_processing(flow_cell_dir: Path, demultiplexed_runs_dir: Pa
         flow_cell_is_ready = False
 
     return flow_cell_is_ready
+
+
+def get_existing_manifest_file(source_directory: Path) -> Path | None:
```
Comment (ChrOertlin) [inline cg/cli/demultiplex/copy_novaseqx_demultiplex_data.py:122]: I just moves these functions into this directory. They will likely need a better place at some point

File: cg/cli/demultiplex/finish.py [hunk 4]
```diff
@@ -17,32 +20,43 @@ def finish_group():
...
-    demux_post_processing_api.set_dry_run(dry_run=dry_run)
-    is_error_raised: bool = demux_post_processing_api.finish_all_flow_cells()
+def post_process_all_illumina_runs(context: CGConfig, dry_run: bool):
+    """Command to post process all demultiplexed Illumina runs."""
+
+    post_processing_service = IlluminaPostProcessingService(
+        status_db=context.status_db,
+        housekeeper_api=context.housekeeper_api,
+        dry_run=dry_run,
+        force=False,
```
Comment (ChrOertlin) [inline cg/cli/demultiplex/finish.py:57]: this force needs actually the be implemented into the new post processing flow...

File: tests/cli/demultiplex/test_verify_syncing.py [hunk 5]
```diff
@@ -106,3 +106,12 @@ def assert_manifest_file_contains_all_files(
...
     assert all(file in files_in_manifest for file in all_files_in_directory)
+
+
+def get_all_files_in_directory_tree(directory: Path) -> list[Path]:
+    """Get the relative paths of all files in a directory and its subdirectories."""
+    files_in_directory: list[Path] = []
+    for subdir, _, files in os.walk(directory):
+        subdir = Path(subdir).relative_to(directory)
+        files_in_directory.extend([Path(subdir, file) for file in files])
+    return files_in_directory
```
Comment (ChrOertlin) [inline tests/cli/demultiplex/test_verify_syncing.py:117]: will move this to utils... it was somewhere weird for the conftests previosuly.

File: tests/services/illumina_services/illumina_metrics_service/parsers/test_illumina_metrics_service.py [hunk 6]
```diff
@@ -1,82 +1 @@
...
-    flow_cell.set_sample_sheet_path_hk(hk_path=sample_sheet_path)
-
-    # WHEN creating undetermined sequencing statistics specifying a lane without undetermined reads
-    metrics: list[SampleLaneSequencingMetrics] = (
-        illumina_metrics_service.create_undetermined_non_pooled_metrics(flow_cell=flow_cell)
-    )
-
-    # THEN an empty list is returned
-    assert not metrics
+"""Test for service dtos."""
```
Comment (ChrOertlin) [inline tests/services/illumina_services/illumina_metrics_service/parsers/test_illumina_metrics_service.py:1]: Need to add tests for DTOS, out of scope here atm.

File: cg/cli/demultiplex/finish.py [hunk 7]
```diff
@@ -1,13 +1,16 @@
 """Commands to finish up after a demultiplex run."""
```
Comment (diitaz93) [inline cg/cli/demultiplex/finish.py:1]:
```suggestion
"""Commands to post-process a demultiplex run."""
```

File: cg/services/illumina_services/illumina_post_processing_service/illumina_post_processing_service.py [hunk 8]
```diff
@@ -31,15 +31,25 @@
...
+        housekeeper_api: HousekeeperAPI,
+        demultiplexed_runs_dir: Path,
+        dry_run: bool,
+        force: bool,
+    ) -> None:
         self.status_db: Store = status_db
         self.hk_api: HousekeeperAPI = housekeeper_api
+        self.demultiplexed_runs_dir = demultiplexed_runs_dir
         self.dry_run: bool = dry_run
+        self.force: bool = force
```
Comment (diitaz93) [inline cg/services/illumina_services/illumina_post_processing_service/illumina_post_processing_service.py:52]: seems that nothing uses force, should we remove it?

File: cg/services/illumina_services/illumina_post_processing_service/illumina_post_processing_service.py [hunk 9]
```diff
@@ -31,15 +31,25 @@
...
+        status_db: Store,
+        housekeeper_api: HousekeeperAPI,
+        demultiplexed_runs_dir: Path,
+        dry_run: bool,
+        force: bool,
+    ) -> None:
         self.status_db: Store = status_db
         self.hk_api: HousekeeperAPI = housekeeper_api
+        self.demultiplexed_runs_dir = demultiplexed_runs_dir
         self.dry_run: bool = dry_run
```
Comment (diitaz93) [inline cg/services/illumina_services/illumina_post_processing_service/illumina_post_processing_service.py:51]: The dry run is only used in the endpoint function and not propagated around, maybe it does not need to be an attibute, just a parameter of `post_process_illumina_flow_cell`

File: tests/fixture_plugins/demultiplex_fixtures/path_fixtures.py [hunk 10]
```diff
@@ -605,3 +634,141 @@ def spring_meta_data_file(spring_dir: Path) -> Path:
 def spring_file_father(spring_dir: Path) -> Path:
     """Return the path to a second existing spring file."""
     return Path(spring_dir, "dummy_run_002.spring")
+
+
+# temporary fastq paths for canonical illumina runs
+def make_tmp_fastq_paths(tmp_path, run_id, sample_ids: List[str]) -> dict[str, Path]:
```
Comment (diitaz93) [inline tests/fixture_plugins/demultiplex_fixtures/path_fixtures.py:640]:
```suggestion
def make_tmp_fastq_paths(tmp_path, run_id, sample_ids: list[str]) -> dict[str, Path]:
```

File: tests/fixture_plugins/demultiplex_fixtures/path_fixtures.py [hunk 11]
```diff
@@ -2,6 +2,7 @@
 
 import shutil
 from pathlib import Path
+from typing import List
```
Comment (diitaz93) [inline tests/fixture_plugins/demultiplex_fixtures/path_fixtures.py:5]: use the python type hints

File: tests/fixture_plugins/demultiplex_fixtures/path_fixtures.py [hunk 12]
```diff
@@ -605,3 +634,141 @@ def spring_meta_data_file(spring_dir: Path) -> Path:
 def spring_file_father(spring_dir: Path) -> Path:
     """Return the path to a second existing spring file."""
     return Path(spring_dir, "dummy_run_002.spring")
+
+
+# temporary fastq paths for canonical illumina runs
+def make_tmp_fastq_paths(tmp_path, run_id, sample_ids: List[str]) -> dict[str, Path]:
+    fastq_paths: dict[str, Path] = {}
+    tmp_path_parent: Path = Path(tmp_path, run_id)
```
Comment (diitaz93) [inline tests/fixture_plugins/demultiplex_fixtures/path_fixtures.py:642]:
```suggestion
    tmp_path_parent = Path(tmp_path, run_id)
```

File: tests/fixture_plugins/demultiplex_fixtures/path_fixtures.py [hunk 13]
```diff
@@ -605,3 +634,141 @@ def spring_meta_data_file(spring_dir: Path) -> Path:
...
+    """Return the path to a temporary NovaSeqX FASTQ file."""
+    return make_tmp_fastq_paths(
+        tmp_illumina_demultiplexed_runs_directory,
+        novaseq_x_flow_cell_full_name,
+        selected_novaseq_x_sample_ids,
+    )
+
+
+@pytest.fixture
+def tmp_fastq_files_for_all_canocial_illumina_demultiplexed_runs(
```
Comment (diitaz93) [inline tests/fixture_plugins/demultiplex_fixtures/path_fixtures.py:750]:
```suggestion
def tmp_fastq_files_for_all_canonical_illumina_demultiplexed_runs(
```

**Review by diitaz93 — APPROVED** [review]
It really looks amazing 🙌 My only suggestion would be to remove the `force` from the post processing service and remove the dry run as class attribute and add it to the service endpoint instead

File: cg/cli/demultiplex/finish.py (hunk 4 above)
Comment (diitaz93) [reply cg/cli/demultiplex/finish.py:57]: ohh ok

File: cg/cli/demultiplex/copy_novaseqx_demultiplex_data.py (hunk 3 above)
Comment (diitaz93) [reply cg/cli/demultiplex/copy_novaseqx_demultiplex_data.py:122]: I agree

File: cg/services/illumina_services/illumina_post_processing_service/illumina_post_processing_service.py (hunk 9 above)
Comment (ChrOertlin) [reply cg/services/illumina_services/illumina_post_processing_service/illumina_post_processing_service.py:51]: will check in new pr

**Pull request merged**